QEfficient/compile/qnn_compiler.py

```diff
@@ -0,0 +1,386 @@
+            for key, value in self.qnn_config[QnnConstants.QNN_COMPILATION_BACKEND_STR].items():
+                qnn_compile_backend[key] = value
+
+        qnn_compile_backend_json_path = os.path.join(self.qpc_path, "qnn_compile_backend.json")
+        create_json(qnn_compile_backend_json_path, qnn_compile_backend)
+        return qnn_compile_backend_json_path
+
+    def create_qnn_compiler_config_json(self) -> str:
+        """
+        Creates qnn_compiler_config.json file containing path to qnn_compile_backend.json file & shared_library_path.
+        Config file is passed to QNN context-binary-generator.
+
+        """
+        qnn_compiler_config = {
+            "backend_extensions": {
+                "config_file_path": self.create_qnn_compile_backend_json(),
+                "shared_library_path": QnnConstants.QNN_CONTEXT_LIB_NET_RUN_EXTENSIONS.format(
+                    self.qnn_sdk_path, self.qnn_target
+                ),
+            }
+        }
+        qnn_compiler_config_json_path = os.path.join(self.qpc_path, "qnn_compiler_config.json")
+        create_json(qnn_compiler_config_json_path, qnn_compiler_config)
+        return qnn_compiler_config_json_path
+
+    def compile(self) -> str:
+        """
+        Compiles the given ``ONNX`` model during object creation using QNN compiler and saves the compiled ``qpc`` package at ``qpc_path``.
+            - Creates convertor command and convert onnx model to model.dlc using qairt-convertor
+            - command line arguments and qnn_config.json (if provided) are used to create qnn_compiler_config.json for context-binary-generator
+            - model.dlc from convertor stage is passed into context-binary-generator command to create programqpc.bin.
+
+        Returns:
+            :str: Path to compiled ``qpc`` package.
+        """
+        if not (
+            self.qnn_config
+            and (QnnConstants.SKIP_QNN_CONVERTOR_STEP_STR in self.qnn_config)
+            and self.qnn_config[QnnConstants.SKIP_QNN_CONVERTOR_STEP_STR]
+        ):
+            converter_cmd = self.converter()
+            execute("convertor", converter_cmd, self.qpc_path)
+
+        if not os.path.isfile(self.dlc_model_path):
+            raise FileNotFoundError(
+                f"file {self.dlc_model_path} needs to exist in the qpc_path{self.qpc_path}. Please rerun infer/compile Api"
+            )
+
+        self.qnn_binary_dir = os.path.join(self.qpc_path, "qpcs")
+        if os.path.isdir(self.qnn_binary_dir):
+            shutil.rmtree(self.qnn_binary_dir)
+        os.makedirs(self.qnn_binary_dir)
+
+        ctx_bin_cmd = self.generate_context_binary()
+        execute("context_binary", ctx_bin_cmd, self.qpc_path)
+
+        print("\n===================== Compilation Done! =====================\n")
+        return self.qnn_binary_dir
+
+    def converter(self):
+        """
+        Creates QNN convertor command using provided options.
+
+        ``IMMUTABLE`` Args:
```

Comment:
Please remove ``IMMUTABLE`` for consistency across codebase. Also include the return statement in doc string